libdirections-onboard/src/test/java/com/mapbox/navigation/route/onboard/MapboxOnboardRouterTest.kt

```diff
@@ -96,6 +98,50 @@ class MapboxOnboardRouterTest {
         verify { routerCallback.onResponse(any()) }
     }
 
+    /**
+     * This test works when run standalone, but when in sequence with other tests it fails because it uses a test dispatcher, even though it's umocked.
+     */
+    @Test
+    fun checkCallbackCalledOnCancel() {
+        unmockkObject(ThreadController)
+        onboardRouter = MapboxOnboardRouter(navigator, MapboxOnboardRouterConfig(TILE_PATH), logger)
+        every { navigator.getRoute(any()) } returns routerResultSuccess
+
+        val latch = CountDownLatch(1)
+        val callback: Router.Callback = object : Router.Callback {
+            override fun onResponse(routes: List<DirectionsRoute>) {
+                Assert.fail()
+            }
+
+            override fun onFailure(throwable: Throwable) {
+                Assert.fail()
+            }
+
+            override fun onCanceled() {
+                latch.countDown()
+            }
+        }
+        onboardRouter.getRoute(routerOptions, callback)
+        onboardRouter.cancel()
+
+        if (!latch.await(5, TimeUnit.SECONDS)) {
+            Assert.fail("onCanceled not called")
+        }
+    }
+
+    /**
+     * This test fails as well because the job canceled before starting. Adding CancelationExceptionHandler to `launch` call doesn't help though.
+     */
+    @Test
+    fun checkCallbackCalledOnCancel2() = coroutineRule.runBlockingTest {
+        every { navigator.getRoute(any()) } returns routerResultSuccess
+        pauseDispatcher()
+        onboardRouter.getRoute(routerOptions, routerCallback)
+        onboardRouter.cancel()
+        resumeDispatcher()
+        verify { routerCallback.onCanceled() }
+    }
+
```

Comment:
@olegzil @korshaknn do you know whether there's a way for us to test coroutine cancelation? All the things that I've come up with failed.

Reply:
Changes introduced in https://github.com/mapbox/mapbox-navigation-android/pull/2434/commits/9cb968602230489aedd1b526b098d7cc309b09f5 look 👌 

I was thinking something around these lines 👇Wouldn't that work too?

```kotlin
fun test = rule.runBlockingTest {
    // logic
    val scope = TestCoroutineScope()
    val job = scope.launch { suspendThatShouldCancel() }
    scope.cancel()

    // logic to check that's been cancelled 
    assertFalse(job.isActive)
}
```

cc @LukasPaczos @korshaknn @olegzil

Reply:
@Guardiola31337 @LukasPaczos @olegzil  please check a new [test](https://github.com/mapbox/mapbox-navigation-android/pull/2434/commits/7f4be490b91d0a9d502ba79708a8efa792c9cc8b) with logic, that @Guardiola31337 provided.
I've made some investigation:
we can't just call
```
val job = scope.launch { suspendThatShouldCancel() }
scope.cancel()
```
1. `suspendThatShouldCancel()` should run on a separate dispatcher not to block main one
2. `suspendThatShouldCancel()` should be suspended for some time to allow main thread cancel it
please check my comments inside the commit.

also we need to add a Job to `TestCoroutineScope`. It will be a parent job for all coroutines launched on that scope. Without that job we can't call `scope.cancel()` after `scope.launch{...}`, we will get 
`java.lang.IllegalStateException: Scope cannot be cancelled because it does not have a job`